pkg/antctl/raw/supportbundle/command.go

```diff
@@ -641,9 +654,157 @@ func processResults(resultMap map[string]error, dir string) error {
 		err = writeFailedNodes(dir, failedNodes)
 	}
 
+	// download logs from kubernetes api
+	if failedNodes != nil || controllerFail {
+		err := downloadPodInfoFromKubernetes(antreaClientset, k8sClient, failedNodes, controllerFail, dir)
+		if err != nil {
+			fmt.Println("Failed to download logs from kubernetes api: " + err.Error())
+		} else {
+			allFailed = false
+		}
+	}
+
 	if allFailed {
 		return fmt.Errorf("no data was collected: %s", genErrorMsg(resultMap))
 	} else {
 		return err
 	}
 }
+
+// downloadLogsFromKubernetes will try to download pod logs from kubernetes api for failed nodes and controller, as well as controllerinfo/agentinfo...
+func downloadPodInfoFromKubernetes(antreaClientset antrea.Interface, k8sClient kubernetes.Interface, failedNodes []string, isControllerFail bool, dir string) error {
+	agentInfoList, err := antreaClientset.CrdV1beta1().AntreaAgentInfos().List(context.TODO(), metav1.ListOptions{ResourceVersion: "0"})
+	if err != nil {
+		return err
+	}
+
+	agentInfoMap := map[string]v1beta1.AntreaAgentInfo{}
+	for _, agentInfo := range agentInfoList.Items {
+		agentInfoMap[agentInfo.Name] = agentInfo
+	}
+
+	pods, err := k8sClient.CoreV1().Pods("kube-system").List(context.TODO(), metav1.ListOptions{
+		ResourceVersion: "0",
+		LabelSelector:   "app=antrea",
+	})
+	if err != nil {
+		return err
+	}
+
+	failedNodesMap := make(map[string]struct{})
+	for _, node := range failedNodes {
+		failedNodesMap[node] = struct{}{}
+	}
+
+	var errors []error
+
+	for _, pod := range pods.Items {
+
+		tmpDir, err := afero.TempDir(defaultFS, "", "bundle_tmp_")
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		defer defaultFS.RemoveAll(tmpDir)
+
+		if pod.Labels["component"] == "antrea-controller" && isControllerFail {
+			controllerInfo, err := antreaClientset.CrdV1beta1().AntreaControllerInfos().Get(context.TODO(), "antrea-controller", metav1.GetOptions{})
+			if err != nil {
+				errors = append(errors, err)
+			}
+			data, err := yaml.Marshal(controllerInfo)
+			if err != nil {
+				errors = append(errors, err)
+			}
+			err = afero.WriteFile(defaultFS, filepath.Join(tmpDir, "controllerinfo"), data, 0644)
+			errors = append(errors, err)
+
+			err = downloadPodLogs(k8sClient, "controller", pod.Namespace, pod.Name, []string{"antrea-controller"}, dir, tmpDir)
+			errors = append(errors, err)
+
+		}
+
+		if _, exist := failedNodesMap[pod.Spec.NodeName]; !exist {
+			continue
+		}
+
+		if pod.Labels["component"] == "antrea-agent" {
+			if agentInfo, ok := agentInfoMap[pod.Spec.NodeName]; ok {
+				data, err := yaml.Marshal(agentInfo)
+				if err != nil {
+					errors = append(errors, err)
+				}
+				err = afero.WriteFile(defaultFS, filepath.Join(tmpDir, "agentinfo"), data, 0644)
+				errors = append(errors, err)
+			}
+
+			var containerNames []string
+			for _, container := range pod.Spec.Containers {
+				containerNames = append(containerNames, container.Name)
+			}
```

Comment:
```suggestion
		        containerNames = append(containerNames, pod.Spec.Containers...)
```

Reply:
@tnqn All updated except this part. sorry i didn't follow.. i checked that a Containers's string format is not it's name